components/collector/src/source_collectors/harbor/security_warnings.py

```diff
@@ -0,0 +1,127 @@
+"""Harbor security warnings collector."""
+
+from abc import ABC
+from urllib.parse import quote, unquote
+
+from base_collectors import SourceCollector
+from collector_utilities.functions import match_string_or_regular_expression
+from collector_utilities.type import URL
+from model import Entities, Entity, SourceResponses
+
+
+class HarborBase(SourceCollector, ABC):
+    """Base class for Harbor collectors."""
+
+    async def _get_source_responses(self, *urls: URL, **kwargs) -> SourceResponses:
+        """Extend to follow Harbor pagination links, if necessary."""
+        all_responses = responses = await super()._get_source_responses(*urls, **kwargs)
+        while next_urls := self._next_urls(responses):
+            # Retrieving consecutive big responses without reading the response hangs the client, see
+            # https://github.com/aio-libs/aiohttp/issues/2217
+            for response in responses:
+                await response.read()
+            all_responses.extend(responses := await super()._get_source_responses(*next_urls, **kwargs))
+        return all_responses
+
+    def _next_urls(self, responses: SourceResponses) -> list[URL]:  # skipcq: PYL-R0201
+        """Return the next (pagination) links from the responses."""
+        return [URL(next_url) for response in responses if (next_url := response.links.get("next", {}).get("url"))]
+
+
+class HarborSecurityWarnings(HarborBase):
+    """Harbor collector for security warnings."""
+
+    async def _api_url(self) -> URL:
+        """Extend to add the Harbor REST API base path."""
+        return URL(await super()._api_url() + "/api/v2.0")
+
+    async def _get_source_responses(self, *urls: URL, **kwargs) -> SourceResponses:
+        """Extend because we need to do multiple requests to get all the data we need."""
+        responses = SourceResponses()
+        projects_api = URL(f"{await self._api_url()}/projects")
+        projects_responses = await super()._get_source_responses(projects_api, **kwargs)
+        for projects_response in projects_responses:
```

Comment:
Is dit niet op te breken in meerdere functies? Misschien maakt dat het leesbaarder.